git-branchless-hook/src/lib.rs

```diff
@@ -289,13 +326,14 @@ mod reference_transaction {
     #[instrument]
     pub fn parse_reference_transaction_line(
         line: &str,
+        resolver: &TransactionReferenceFinder,
     ) -> eyre::Result<ParsedReferenceTransactionLine> {
         let fields = line.split(' ').collect_vec();
         match fields.as_slice() {
             [old_value, new_value, ref_name] => Ok(ParsedReferenceTransactionLine {
                 ref_name: ReferenceName::from(*ref_name),
-                old_oid: MaybeZeroOid::from_str(old_value)?,
-                new_oid: MaybeZeroOid::from_str(new_value)?,
+                old_oid: resolver.oid_for_transaction_reference(old_value)?,
```

Comment:
issue: This is only a function that claims to "parse" the line, and isn't (currently) supposed to do any processing/resolution. The question is whether we would prefer to preserve that, or just concede that mixing the parsing and resolution steps together is somehow better/more maintainable.

Based on my previous experience, I think it's actually best to keep this as a pure parsing function and handle this in a later phase. Specifically, it turned out to be valuable to separate and test the parsing of `git status` output in isolation.

To fix this, let's return some kind of `enum` and preserve the symbolically-represented input, rather than resolving it immediately. It looks like there already exists the `git::ReferenceTarget` type which we can use here.

- If it's not suitable for some reason, you can make your own, scoped to this file or even a function.
- I don't know your level of Rust experience. Here's an example of constructing the `Cow` if you need it (I recall it wasn't obvious to me): https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=afa8cdac4b58b09b217449b1d4e0897a. Feel free to clone any borrowed `str`s to owned `String`s if it helps.

After that, I think we can resolve the reference OIDs when constructing `events: Vec<Events>`. (Also depending on your level of Rust experience, you may find it easier to use an explicit loop rather than `into_iter`; up to you.)

suggestion (non-blocking): It may make sense to move the parsing code to belong to `ReferenceTarget`, either as an implementation of `FromStr`, or by a custom function. I don't know if any other caller will ever need it, though, so fine to keep local if you prefer.

comment: I assume that the field names in `ParsedReferenceTransactionLine` were meant to match the `reference-transaction` hook documentation directly, so the appropriate new names would be `old_value` and `new_value`. (I think we should actually reorder the fields to match the input order as well, so `ref_name` would be last, but not a big deal.)